active/0003-extensible-resource-deployment-observability-improvements.md

```diff
@@ -0,0 +1,249 @@
+
+#### Adding `deploymentProviders`
+
+The `Microsoft.Resources/deployments` API will be updated to include a `deploymentProviders` array in both the PUT and GET response bodies. This modification is intended to incorporate information about deployment providers that are involved in executing the parent deployment and nested deployments to offer a more comprehensive deployment view.
+
+> It's worth noting that we cannot use the name `providers` as the property `providers` is already assigned to return Azure Resource Provider information. While an alternative approach could involve renaming the current `providers` property to `resourceProviders` for enhanced accuracy, the solution should be avoided as it constitutes a breaking change.
+
+```diff
+{
+  ...
+  "type": "Microsoft.Resources/deployments",
+  "properties": {
+    ...
+    "providers": [
+      {
+        "namespace": "Microsoft.Storage",
+        "resourceTypes": [
+          {
+            "resourceType": "storageAccounts",
+            "locations": [
+              "eastus"
+            ]
+          }
+        ]
+      }
+    ],
++   "deploymentProviders": [ // Can we come up with a better name?
++     {
+        // deploymentId is used to uniquely identify a provider since a nested deployment could import the same provider with the same alias.
++       "deploymentId": "/subscriptions/00000000-0000-0000-0000-000000000001/resourceGroups/my-resource-group/providers/Microsoft.Resources/deployments/my-deployment",
++       "alias": "az",
++       "name": "AzureResourceManager",
++       "version": "1.0.0"
++     }  ,
++     {
++       "deploymentId": "/subscriptions/00000000-0000-0000-0000-000000000001/resourceGroups/my-resource-group/providers/Microsoft.Resources/deployments/my-deployment",
++       "alias": "kubernetes",
++       "name": "Kubernetes",
++       "version": "1.27.8"
++     }
+    ]
+  }
+}
+```
+
+#### Updating the `outputResources` to include more identifiable information
+
+To improve the current structure where each `outputResources` object only has a resource `id`, it's essential to add more properties for identifying extensible resources, which lack resource IDs. These additions should include the symbolic name, the resource type, the ID of the deployment containing the resource, and the deployment provider alias.
+
+```diff
+{
+  ...
+  "type": "Microsoft.Resources/deployments",
+  "properties": {
+    "outputResources": [
+      // Azure resource
+      {
+        "id": "/subscriptions/00000000-0000-0000-0000-000000000001/resourceGroups/my-resource-group/providers/Microsoft.Storage/storageAccounts/my-storage-account"
+      },
+      // Extensible resource
+      {
++       "symbolicName": "myService",
++       "resourceType": "core/Service@v1",
++       "deploymentId": "/subscriptions/00000000-0000-0000-0000-000000000001/resourceGroups/my-resource-group/providers/Microsoft.Resources/deployments/my-deployment",
++       "deploymentProvider": "kubernetes"
+      }
```

Comment:
How would a user or tool identify the actual resource? This is the difference between:

- *A Kubernetes `Deployment`* was created
- *A Kubernetes `Deployment` called `my-deployment` in namespace `my-namespace` was created

I don't think the former is super useful. The symbolic name doesn't help because it refers to the implementation details of the template (which can change).

Reply:
For users, `deploymentId`` and `symbolicName`` should suffice to link a resource to its Bicep module and its declaration within that module, assuming the template remains unchanged. However, for services like Deployment Stacks that need to uniquely identify a resource even when the template changes, a unique ID for extensible resources is necessary. This concept is further explored in the subsequent REP: https://github.com/Azure/bicep-reps/pull/4.